File: swift/PraUTS/PraUTS/OperatingSystem.swift
//
//  OperatingSystem.swift
//  PraUTS
//
//  Created by Dodi Aditya on 06/11/23.
//

import Foundation

class OperatingSystem {
    private(set) var name: String
    private(set) var version: String
    private(set) var releaseDate: Date
    private(set) var developer: String
    private(set) var language: String
    private(set) var serialNumber: String
    private var isOn: Bool
    
    init(name: String, version: String, releaseDate: Date, developer: String, language: String, serialNumber: String) {
        self.name = name
        self.version = version
        self.releaseDate = releaseDate
        self.developer = developer
        self.language = language
        self.serialNumber = serialNumber
        
        self.isOn = true
    }
    
    convenience init(name: String, version: String, releaseDate: Date, developer: String, language: String) {
        self.init(name: name, version: version, releaseDate: releaseDate, developer: developer, language: language, serialNumber: "")
    }
    
    func printInfo() {
        print("OPERATING SYSTEM")
        print("Name            : \(name)");
        print("Version         : \(version)");
        print("Release         : \(getFormattedReleaseDate())");
        print("Developer       : \(developer)");
        print("Language        : \(language)");
        print("Serial Number   : \(serialNumber)");
    }
    
    func boot() {
        if !isOn {
            print("Booting")
            self.isOn = true
        } else {
            print("System is already ON.")
        }
    }
    
    func shutdown() {
        if (isOn) {
            print("Logging off this system....");
            isOn = false;
        } else {
            print("This system is currently OFF.");
        }
    }
    
    func restart() {
        if (isOn) {
            print("Restarting...")
            shutdown()
            DispatchQueue.main.asyncAfter(deadline: .now() + 3) {
                self.boot()
            }
        }
    }
    
    func getFormattedReleaseDate() -> String {
        releaseDate.formatted(date: .long, time: .omitted)
    }
    
    func setLanguage(language: String) {
        self.language = language
    }
    
    func getBootStatus() -> String {
        "This system is currently \(isOn ? "ON" : "OFF")"
    }
}

File: java/PraUTS/src/main/java/PraUTS/OperatingSystem.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package PraUTS;

import java.text.SimpleDateFormat;
import java.util.Date;

/**
 *
 * @author dodiaditya
 */
public class OperatingSystem {
    private String name;
    private String version;
    private Date release_date;
    private String developer;
    private String language;
    private String serial_number;
    private boolean isOn = false;
    
    public OperatingSystem(String name, String version, Date release_date, String developer, String language) {
        this.name = name;
        this.version = version;
        this.release_date = release_date;
        this.developer = developer;
        this.language = language;
        boot();
    }
    
    public OperatingSystem(String name, String version, Date release_date, String developer, String language, String serial_number) {
        this(name, version, release_date, developer, language);
        this.serial_number = serial_number;
    }
    
    public void printInfo() {
        System.out.println("OPERATING SYSTEM");
        System.out.println("Name            : " + this.name);
        System.out.println("Version         : " + this.version);
        System.out.println("Release         : " + this.getFormattedReleaseDate());
        System.out.println("Developer       : " + this.developer);
        System.out.println("Language        : " + this.language);
        System.out.println("Serial Number   : " + this.serial_number);
        
        printAdditionalInfo();
    }
    
    public void printAdditionalInfo() { }
    
    public void updateVersion(String version) {
        System.out.println("Updating version....");
        this.version = version;
    }
    
    public void shutdown() {
        if (this.isOn) {
            System.out.println("Logging off this system....");
            this.isOn = false;
        } else {
            System.out.println("This system is currently off.");
        }
    }
    
     public void boot() {
        if (!this.isOn) {
            System.out.println("Booting....");
            this.isOn = true;
        } else {
            System.out.println("This system is currently on.");
        }
    }
     
    public void restart() {
        if (this.isOn) {
            System.out.println("Restarting...");
            shutdown();
            try {
                Thread.sleep(3000);
                boot();
            } catch(InterruptedException e) {
                e.printStackTrace();
            }
        } 
    }
    
    public String getName() {
        return this.name;
    }
    
    public String getVersion() {
        return this.version;
    }
    
    public Date getReleaseDate() {
        return this.release_date;
    }
    
    public String getFormattedReleaseDate() {
        SimpleDateFormat dateFormat = new SimpleDateFormat("dd MMMM yyyy");
        return dateFormat.format(this.release_date);
    }
    
    public String getDeveloper() {
        return this.developer;
    }
    
    public String getLanguage() {
        return this.language;
    }
    
    public void setLanguage(String lang) {
        this.language = lang;
    }
    
    public String getSerialNumber() {
        return this.serial_number;
    }
    
    public String getBootStatus() {
        String msg = "This system is currently ";
        
        if (this.isOn) {
            msg += "ON";
        } else {
            msg += "OFF";
        }
        
        return msg;
    }
}
